### ISSUE TITLE
dotnet muxer reports JSON syntax error when global.json is saved in UTF8+BOM

### ISSUE THREAD

**Issue by pakrym** [dotnet/cli#2159]
## Steps to reproduce
1. Save global.json in utf8 with byte order mark.
2. dotnet build

## Expected  behavior
dotnet build shows no errors

## Actual behavior
```
A JSON parsing exception occurred: * Line 1, Column 2 Syntax error: Malformed token
```

## Environment data
`dotnet --version` output:
```
A JSON parsing exception occurred: * Line 1, Column 2 Syntax error: Malformed token
1.0.0-beta-002171
```

**Comment by pakrym** [comment]
//cc @piotrpMSFT @davidfowl @victorhurdugaci @pranavkm

**Comment by pakrym** [comment]
This makes build to fail.
```
dotnet build
A JSON parsing exception occurred: * Line 1, Column 2 Syntax error: Malformed token
Project 1 (.NETStandardApp,Version=v1.5) will be compiled because the CLI changed
Compiling 1 for .NETStandardApp,Version=v1.5
D:\temp\1\A JSON parsing exception occurred: * Line 1, Column 2 Syntax error: Malformed token

Compilation failed.
    0 Warning(s)
    1 Error(s)

Time elapsed 00:00:05.4565174
```

**Comment by gkhanna79** [comment]
@pakrym Why is this blocking given that JSON parsing was working all this while?

**Comment by pakrym** [comment]
Because now its output makes `dotnet build` to fail

**Comment by gkhanna79** [comment]
Have you changed the format in which the .json files are saved?

**Comment by pakrym** [comment]
No, it reports false positive for files with BOM

**Comment by gkhanna79** [comment]
> Save global.json in utf8 with byte order mark.

How was this saved earlier when this worked? I am not aware of any JSON parsing changes that would result in this.

**Comment by pakrym** [comment]
Reporting changed in https://github.com/dotnet/cli/commit/1428eb81cfb1e81251b8241d08e5fefd95882153, it started printing line numbers into `stderr` and `dotnet build` thinks its `csc` errors.

**Comment by gkhanna79** [comment]
Are you suggesting that if trace_error was not executed, the build would have succeeded as that would be strange given that json parsing failed?

**Comment by pakrym** [comment]
Global.json is perfectly valid json file, it just has UTF8 BOM encoding which `corehost` can not process correctly. So yes, build would have succeeded if this error wasn't written to `stderr`.

**Comment by pakrym** [comment]
I'm not fixing it

**Comment by pakrym** [comment]
This will only make it write to `stdout` instead `stderror`

**Comment by gkhanna79** [comment]
Casablanca does not support BOM. If you think this is blocking for RTM, let us discuss.

**Comment by pakrym** [comment]
I think it blocks rc2, VS saves json files with BOM by default.

**Comment by gkhanna79** [comment]
With your workaround in place, I dont believe this is blocking for RC2. We can look into what can be done for RTM for Casablanca.

**Comment by pakrym** [comment]
All users would see error message that is not true, how is this not an issue?

**Comment by pakrym** [comment]
And sdk selection feature won't work at all.

**Comment by gkhanna79** [comment]
Are there other editors that allow you to save JSON without BOM marker? If so, can that not be a workaround until RTM to author the json file? Or, are you suggesting there is a case where VS is used to create a project type that will have .json files with BOM markers?

**Comment by pakrym** [comment]
`File->Create project->Core console application` creates `global.json` with BOM

**Comment by schellap** [comment]
Cc @BillHiebert

**Comment by davidfowl** [comment]
Did anyone look at a fix?

**Comment by gkhanna79** [comment]
I am looking at what Casablanca supports and what we can do.

**Comment by schellap** [comment]
@gkhanna79 we should just istream::get and match for the BOM and istream::gseek to 0 if we don't find it. We shouldn't probably mess with Casablanca. `global.json` is a little different from `deps.json` in that it is conformant to the platform spec for such files.

**Comment by davidfowl** [comment]
We can start a thread with that team and work around the BOM issue hopefully.

**Comment by gkhanna79** [comment]
They have some stuff done, to handle bom, in their httphelpers implementation:

https://github.com/Microsoft/cpprestsdk/blob/1e487cc2d158f1cc4a4627beaf8d0b7663957b8b/Release/src/http/common/http_helpers.cpp

**Comment by gkhanna79** [comment]
Fixed with https://github.com/dotnet/cli/pull/2186.

**Comment by naamunds** [comment]
@gkhanna79 I just ran into this issue with a global.json that has a UTF-16 BOM.

I'm testing .NET on a Windows Nano Server container, so I used PowerShell's Out-File cmdlet to create the global.json file, which appears to use UTF-16 by default.

I was able to work around the issue by removing the BOM by using `[IO.File]::WriteAllText`.  I also verified that using a UTF-8 BOM is another workaround, e.g. by using `Out-File -Encoding utf8`.

I'm wondering if it would be worth fixing this issue for UTF-16 in addition to UTF-8, because I think people might run into it, especially because the error message does not make it obvious what the problem is.

cc: @dleeapho @MichaelSimons
